### PR TITLE
Fix ValueGenerated On Add Or Update, , make ValueGeneratedOnUpdate() and ValueGeneratedOnAddOrUpdate() working

### PR DESCRIPTION
**Pull request by eation**
Fixes [#6999](https://github.com/dotnet/efcore/issues/6999) and  [#19765](https://github.com/dotnet/efcore/issues/19765) 
docs [#3165](https://github.com/dotnet/EntityFramework.Docs/issues/3165)

When useing ValueGeneratedOnUpdate(), set PropertyBuilder.Metadata.SetAfterSaveBehavior(PropertySaveBehavior.Save) for update.

When useing ValueGeneratedOnAddOrUpdate(), set PropertyBuilder.Metadata.SetBeforeSaveBehavior(PropertySaveBehavior.Save) for add and PropertyBuilder.Metadata.SetAfterSaveBehavior(PropertySaveBehavior.Save) for update.

### CODE REVIEW

**Comment by eation** [comment]
@ajcvickers @AndriySvyryd
Can Cosmos modify explicitly or default PK value with a temporary value on state changed?

https://github.com/dotnet/efcore/blob/50f8b8467b89d874da201a8e3e40149dd2e44a82/test/EFCore.Cosmos.FunctionalTests/Internal/IdentifierShadowValuePresenceTest.cs#L35-L57

Related [issues](https://github.com/dotnet/efcore/issues/15289) and [pull requests](https://github.com/dotnet/efcore/pull/22149)

**Comment by AndriySvyryd** [comment]
#4073 (and #19135) would need to be fixed first

File: test/EFCore.Tests/DbContextTrackingTest.cs [hunk 1]
```diff
@@ -666,12 +666,12 @@ public async Task Can_add_new_entities_to_context_with_key_generation_graph(bool
...
-            context.Entry(gu1).State = EntityState.Unchanged;
-            Assert.Equal(default, gu1.Id);
-            Assert.Equal(EntityState.Unchanged, context.Entry(gu1).State);
-        }
+            if (attachFirst)
+            {
+                context.Entry(gu1).State = EntityState.Unchanged;
+                Assert.NotEqual(default, gu1.Id);
+                Assert.Equal(EntityState.Unchanged, context.Entry(gu1).State);
+            }
```
Comment (Saibamen) [inline test/EFCore.Tests/DbContextTrackingTest.cs:674]: Wrong formatting

File: test/EFCore.Tests/ChangeTracking/Internal/InternalEntryEntrySubscriberTest.cs [hunk 2]
```diff
@@ -577,11 +578,12 @@ private class FullNotificationEntity : INotifyPropertyChanging, INotifyPropertyC
...
-        {
-            get => _id;
-            set => SetWithNotify(value, ref _id);
-        }
+            [DatabaseGenerated(DatabaseGeneratedOption.None)]
+            public int Id
+            {
+                get => _id;
+                set => SetWithNotify(value, ref _id);
+            }
```
Comment (Saibamen) [inline test/EFCore.Tests/ChangeTracking/Internal/InternalEntryEntrySubscriberTest.cs:586]: Wrong formatting

File: src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs [hunk 3]
```diff
@@ -1883,7 +1915,7 @@ public EntityEntry ToEntityEntry()
                     break;
                 case NotifyCollectionChangedAction.Reset:
                     throw new InvalidOperationException(CoreStrings.ResetNotSupported);
-                // Note: ignoring Move since index not important
+                    // Note: ignoring Move since index not important
```
Comment (Saibamen) [inline src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:1918]: Wrong formatting?

Comment (eation) [reply src/EFCore/ChangeTracking/Internal/InternalEntityEntry.cs:1918]: @Saibamen Not sure,but this src file auto-format by vs after last marge

**Comment by ajcvickers** [comment]
As @AndriySvyryd mentioned above, making this kind of change is dangerous without more generally updating the handling of modified key properties, as tracked by #4073 and #19135, and possibly other issues. This kind of change would also need significant new testing to make sure it does not result in bad state that cannot be handled by other parts of the stack. For these reasons, this is not a change we plan to merge.

**Pull request closed without merging**